### DISCUSSION TITLE
Tests don't run queries

### DISCUSSION THREAD

**Discussion by tradical-allstripes** [prisma/prisma#6917] [Q&A] [answered]
Hello Prisma community!

I'm currently trying to write some integration tests that rely on an actual Prisma database, and running into some issues.

First question:
Do you think that using an actual DB is the way to go for this, or is something along the lines of mocking a better idea? I saw [prisma-test-utils](https://github.com/prisma/prisma-test-utils), but it seemed really strange...I couldn't find many usages of it, and absolutely no usages of the code it generates in GitHub. There were fewer total usages of the text `prisma-test-utils` in GitHub than there are stars on the repo! LMK if someone knows more about what's going on here.

Second question:
The issue I'm encountering now is that Prisma is straight up not running the queries in tests, but does not fail or error out in any way. When I rely on lazy connection, providing an invalid DB URL does not even cause it to break (adding a call to `$connect` causes it to fail in that specific case, as expected, but queries still do not run when given the correct DB URL). I have turned on query logging, but nothing prints outside of my `console.log` statements. Just to be sure, I've also been checking the actual table with `psql` after every run; always 0 rows. Here is a version of my prototype test code that has confidential information removed:

```typescript
import { PrismaClient } from '@prisma/client'

// Neither of these do anything. They don't even break if I make the DB URL incorrect

describe('<redacted>Resolver', () => {
  const prisma = new PrismaClient({ log: ['query'] })

  beforeAll(async () => {
    await prisma.$connect()
  })

  afterAll(async () => {
    await prisma.$disconnect()
  })
  test('test create', () => {
    console.log(`CREATE: ${JSON.stringify(prisma.<redacted>.create({
      data: {
        objectId: 'test',
      },
    }))}`)
  })

  test('test find many', async () => {
    console.log(`RESULT: ${JSON.stringify(prisma.<redacted>.findMany({}))}`)
  })
})
```

Here is the output:

```
  console.log
    CREATE: {}

      at Object.<anonymous> (src/__tests__/resolvers/<redacted>Resolver.test.ts:16:13)

  console.log
    RESULT: {}

      at src/__tests__/resolvers/<redacted>Resolver.test.ts:24:13
```

Here is my `datasource` block in `schema.prisma`:

```
datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}
```

Lastly, here is the `DATABASE_URL` field in my `.env` file:

```
DATABASE_URL="postgresql://tradical@localhost:5433/api_tests"
```

Note that I have tried all sorts of restarting and reinstalling `node_modules`

**Comment by janpio** [comment]
> I saw [prisma-test-utils](https://github.com/prisma/prisma-test-utils), but it seemed really strange...I couldn't find many usages of it, and absolutely no usages of the code it generates in GitHub. There were fewer total usages of the text `prisma-test-utils` in GitHub than there are stars on the repo! LMK if someone knows more about what's going on here.

Those were utils to test Prisma - so you are correct that this is _very_ strange in the context you were looking. You made the right choice to ignore it.

**Comment by janpio** [comment, marked as answer]
> The issue I'm encountering now is that Prisma is straight up not running the queries in tests, but does not fail or error out in any way.

Prisma queries are async and need to be awaited. Right now you are not doing that before logging the results - which at that point in time do not exist. Just putting the queries like this probably does not even start them. If you sprinkle in a few `await` or `.then()` the queries should actually be executed.

**Reply by tradical-allstripes** [reply]
Oooooh interesting, I'll give it a try. Thank you for the prompt reply!

Now my only question is "why didn't TS warn me about this?"

**Reply by tradical-allstripes** [reply]
For posterity, I can confirm that @janpio's suggestion works

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
